torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -374,27 +382,125 @@ def __init__(
+
+        # resume from ckpt
+        self._recipe_checkpoint = get_recipe_checkpoint_path(
+            output_dir=self._output_dir,
+            recipe_checkpoint=recipe_checkpoint,
+            resume_from_checkpoint=self._resume_from_checkpoint,
+        )
+
+        self._adapter_checkpoint = get_adapter_checkpoint_path(
+            output_dir=self._output_dir,
+            adapter_checkpoint=adapter_checkpoint,
+            resume_from_checkpoint=self._resume_from_checkpoint,
+            pattern=r"^epoch_(\d+)",
+        )
+
+        # e.g.checkpoint_files = [model-00001-of-00002.safetensors, model-00002-of-00002.safetensors]
+        # Case 1: no resuming from ckpt. Just load from checkpoint_dir.
+        if not self._resume_from_checkpoint:
+            self._checkpoint_paths = self._validate_hf_checkpoint_files(
+                checkpoint_files,
+                input_dir=self._checkpoint_dir,
+                missing_ok=False,
+            )
+
+        # Case 2: Resuming from ckpt, but its full finetuning (no adapter). Just load from output_dir.
+        elif self._adapter_checkpoint is None:
+            self._checkpoint_paths = self._validate_hf_checkpoint_files(
+                checkpoint_files,
+                input_dir=self._output_dir,
+                missing_ok=False,
+            )
+
+        # Case 3: Resuming from ckpt and has an adapter. If save_adapter_only, load from ckpt_dir. Else, load from output_dir.
+        # TODO: we don't have access to "save_adapter_only" flag. So we look at output_dir first, and then ckpt_dir last.
+        else:
+            # Case 3.1: Tentative check {output_dir}/{file_path}. missing_ok=True.
+
+            # TODO: it is annoying that the user has to specify the new file names as "epoch_2/ft-model-00001-of-00001.safetensors",
+            # for example. Ideally, it should just look for files in the latest epoch folder..
+            self._checkpoint_paths = self._validate_hf_checkpoint_files(
+                checkpoint_files,
+                input_dir=self._output_dir,
+                missing_ok=True,
+            )
+
+            self._checkpoint_paths = [
+                file_path for file_path in self._checkpoint_paths if file_path.is_file()
+            ]
+
+            # Case 3.2: Check {checkpoint_dir}/{file_path}
+            if len(self._checkpoint_paths) != len(checkpoint_files):
+                logger.warning(
+                    f"When resuming from ckpt, we could not find all model files in {self._output_dir=}. "
+                    "This is expected if you set `save_adapter_weights_only=True`. In this case, we will load from checkpoint_dir. "
+                    "However, if you set `save_adapter_weights_only=False`, this is unexpected. "
+                    "Perhaps you forgot to add `epoch_{epoch}/` to your filename? "
+                    "Using checkpoint_dir instead..."
+                )
```

Comment:
i don't really like this. The other options are 1) to actually fix the issue, which is knowing save_adapter_only, or 2) silently let it happen, which is dangerous if the adapter + model was trained (e.g. embeddings), and the user forgot to change the file names.

Not sure if there is a 3rd

Reply:
I'm not sure I fully follow this, but for (1) couldn't it be done by just e.g. saving `save_adapter_weights_only` as part of the recipe state and pulling it from there?

Reply:
are you ok with saving it as part of the recipe_state? that would work.

>I'm not sure I fully follow this,
1. User trains a model that has both lora + finetuning (e.g. vision model)
2. User resumes from ckpt, and forgets to update the ckpt files
3. Since we are looking at the ckpt_dir, we will get the **untrained model**, which is a silent bug.

Reply:
removed this. when lora=True, it will always get it from ckpt_dir (which is our current state anyway). Need a follow up PR to address it.

Reply:
Seems like there are a few follow-ups on this PR.. can we file an issue so we can track the different todos in a single place?